<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<title>Supplementary Information, Tolonen et al. 2006</title>
<style type="text/css">
	td { font-size: 9px }
</style>
</head>
<body>
    <h1>Tolonen <i><i>et al.</i></i> 2006:  Gene Set Enrichment Analysis</h1>
    <p>These PowerPoint files contain data relevant to the use of GSEA to detect overrepresentation of genes with NtcA motifs in their promoters among significantly differentially-expressed genes in the Nitrogen starvation time series. The files differ in the stringency and selection criteria used to identify genes with NtcA sites.</p> 
    <p><a href="GSEA_goldenspike_15_15.ppt">GSEA_goldenspike_15_15.ppt</a> defined 'genes with NtcA sites' as genes with NtcA site scores that were among the top 15 for genes with orthologs, plus the genes with the top 15 scoring NtcA sites among genes without orthologs.  This is the way we identified genes with NtcA sites in our article.</p>
    <p><a href="GSEA_goldenspike_20_20.ppt">GSEA_goldenspike_20_20.ppt</a> and <a href="GSEA_goldenspike_ortho_15.ppt">GSEA_goldenspike_ortho_15.ppt</a> were alternative analyses that used different definitions of 'genes with NtcA sites'. Although we did not feature them in our article, we took them into account there in correcting for multiple hypotheses in our Table 2 and provide them here for completeness.</p>
    <p><a href="GSEA_goldenspike_repressed.ppt">GSEA_goldenspike_repressed.ppt</a> is a parallel analysis to search for enrichment of genes with NtcA sites among genes repressed during N starvation.  In these analyses, two sets of genes were compared to the N starvation expression profiles using GSEA: genes with NtcA sites lacking -10 boxes, and genes with NtcA sites that were within 15 bp of a downstream -10 box.  Neither of these analyses supported significant enrichment of genes with putative NtcA sites among genes repressed during N starvation.  

    <p><a href="gsea_pvalues.html">This table</a> contains the P values for enrichment of genes with NtcA sites among genes ordered by significant upregulation for the complete set of analyses described above.  It is a superset of the results presented in Table 2 of our article.</p>
    <br>
    <p><a href="gsea2.m">gsea2.m</a> is the MATLAB function used to perform the analysis. Comments at the beginning of this function describe parameters and their usage, and are also available by typing "help gsea2" from the MATLAB command line.</p>
    <p><a href="NtcA_gsea_tests.tar.gz">Ntca_gsea_tests.tar.gz</a> is a compressed archive that contains the MATLAB script used to perform the GSEA analyses, and all of the numerical data generated by these analyses.  To access the individual files in this file, one must decompress it with gunzip and then untar it with tar -xvf.</p>
    <p>For a more detailed description of all the files shown above, please see <a href="GSEA_README.txt">GSEA_README.txt</a>.</p>
    <br />
    <hr>    
    Latest Update: 13-Feb-2006 by Andy Tolonen.
</body>
</html>
