<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<title>Supplementary Information, Tolonen et al. 2006</title>
</head>
<body>
    <h1>Tolonen <i>et al.</i> 2006:  Supplementary Information</h1>
    <p>This page contains the supporting information for:<br>
	<center>
	<table cellspacing="0" cellpadding="6" border="1" bgcolor="#FFFFFF" width="70%">
	<tr><td>
	<p><b>Global gene expression of Prochlorococcus ecotypes in response to changes in nitrogen availability</b></p>
	<p>Andrew C. Tolonen, John Aach, Debbie Lindell, Zackary I. Johnson, Trent Rector, Robert Steen, George M. Church, Sallie W. Chisholm</p>
	<p>submitted to Moleculary Systems Biology on 15 feb 2006</p>
	</td></tr>
	</table>
	</center>

    <h2>Contents</h2>
    <ul>
      <li><a href="#Goldenspike">Goldenspike Output Files</a></li>
      <li><a href="#Expression">Summary of Goldenspike results: identification of differentially expressed genes</a></li>
      <li><a href="#Kmeans">Kmeans Clustering</a></li>
      <li><a href="#GSEA">Gene Set Enrichment Analysis</a></li>
      <li><a href="#Operons">Validation of operon predictions: correlations with gene expression patterns</a></li>
      <li><a href="#Motifs">Regulatory Motif Finding</a></li>
    </ul>
    <br><br><br>
    <a name="Goldenspike"></a>
    <h2>Goldenspike Output Files</h2>
    <p>Expression summaries were computed from the probe intensities in Affymetrix .CEL files using the Goldenspike R package freely available <a href="http://www.elwood9.net/spike/">here</a>.  Here are the text-formatted output files of the Goldenspike Program for the <a href="Goldenspike/MED4/index.html">MED4</a> and <a href="Goldenspike/MIT9313/index.html">MIT9313</a> N source and N deprivation experiments.</p>
    <p>References:<br> 
    Choe, SE., Boutros, M., Michelson, AM., Church, GM., Halfon, MS. (2005). <a href="http://www.pubmedcentral.gov/articlerender.fcgi?tool=pubmed&pubmedid=15693945">Preferred analysis methods for Affymetrix GeneChips revealed by a wholly defined control dataset.</a> Genome Biology 6: R16.</p>
    <br>
    <a name="Expression"></a>
    <h2>Summary of Goldenspike results: Identification of differentially expressed genes</h2>
    <p>These Excel-formatted compilations of the Goldenspike output files for <a href="Expression/medExpression.xls">MED4</a> and <a href="Expression/mitExpression.xls">MIT9313</a> show the expression of each gene at each time point of the N deprivation time series and on alternative N sources. The first column is the gene name, the next six columns are the log2(-N/+NH4) expression summaries at the 6 timepoints of the N deprivation experiment. The next two columns show log2(alternate N source/ammonium) expression summaries in log-phase growth. The final column shows the gene annotation.</p>
    <p>Differentially-expressed genes between treatments were identified using q-values computed by Goldenspike. The q-value represents the false discovery rate of differentially expressed genes as the fraction of false positives in a group of genes exceeding a statistical cutoff. Differentially expressed genes were defined for K-means clustering, GSEA, and motif finding as q&lt;0.01.  In the above Excel spreadsheets, expression values with q&lt;0.01 have double asterisks (**) and those where 0.01&lt;q&lt;0.05 have single asterisks (*).</p> 
    <p>Note: the Goldenspike output files reflect the Affymetrix-specific gene names used on the chip. <a href="Goldenspike/master_gene_table.txt">master_gene_table.txt</a> is a table correlating the Affymetrix gene names with the common gene names used in this paper.</p> 
    <br> 
    <a name="Kmeans"></a> 
    <h2>Kmeans Clustering</h2>
    <p>The K-means clusters provide a framework to understand the dynamics of the <i>Prochlorococcus</i> response to N stress by identifying genes that responded similarly along the N starvation time course, and may thus be co-regulated.</p> 
    <p>Here are the gene members of each <a href="Kmeans/medClusters.txt">MED4</a> (K=9) and <a href="Kmeans/mitClusters.txt">MIT9313</a> (K=7) cluster described in Results and Discussion. These files contain the gene name, gene function, and Cyanobase functional category for each gene in each cluster.</p> 
    <p>To cluster the expression profiles, the log2 (-N/+NH4) expression summaries generated by Goldenspike clustered were to minimize the sum of the squared Euclidian distances with the 'replicates' option set to 500. Only genes which were differentially expressed (q<0.05) at one or more time point were clustered, representing a total of 410 MED4 genes and 559 MIT9313 genes. The numbers of clusters (K) was chosen using an adaptation of the method of Gibbons and Roth (2002) to maximize the mutual information (MI) between cluster membership and gene functional categories. Gene functional categories were as defined by the <a href="http://www.kazusa.or.jp/cyano/">Cyanobase database</a>.</p> 
    <p>After clustering the genes, we calculated the fraction of genes in each cluster that were in each the Cyanobase functional categories for <a href="./Kmeans/medFractions_all.txt">MED4</a> and <a href="./Kmeans/mitFractions_all.txt">MIT9313</a>.  
In addition, we calculated the fraction of genes for each cluster in each functional category for <a href="./Kmeans/medFractions.txt">MED4</a> and <a href="./Kmeans/mitFractions.txt">MIT9313</a> where the Cyanobase categories "Other", "Hypothetical", and categories represented by fewer than 10 genes were excluded. 

Based on this subset of the functional categories, the statistical significance of the enrichment of individual clusters for each of the functional categories was evaluated in term of hypergeometric P-values (<a href="Kmeans/clusterpvalues.html">click here for details</a>).</p>  
    <p>In order to elucidate the function of unknown genes, we examined the category of the individual genes in the clusters with significant functional category enrichment. Because these clusters were highly enriched for a single functional category, genes in other functional categories within these clusters likely have roles related to the dominant functional category. For <a href="Kmeans/medSigClusters.xls">MED4</a>, these are clusters 7,8,9 and for <a href="Kmeans/mitSigClusters.xls">MIT9313</a> these are clusters 7,8.</p>
<p><a href="./Kmeans/clustering.pdf">Here</a> is a complete description of the methods and results in .pdf file format.</p>
    <p>References:<br> 
    Gibbons, FD and Roth, FP (2002). <a href="http://www.genome.org/cgi/content/abstract/12/10/1574?maxtoshow=&HITS=10&hits=10&RESULTFORMAT=1&author1=gibbons+&author2=roth&andorexacttitle=and&andorexacttitleabs=and&andorexactfulltext=and&searchid=1&FIRSTINDEX=0&sortspec=relevance&resourcetype=HWCIT">Judging the quality of gene expression-based clustering methods using gene annotation.</a>Genome Research. Oct, 12: 1574-1581</p>
    <br> 
    <a name="GSEA"></a> 
    <h2>Gene Set Enrichment Analysis</h2>
    <p>Gene Set Enrichment Analysis (GSEA) is method to evaluate if an <i>a priori</i> defined set of genes correlate with those genes differentially expressed in a microarray dataset. We used GSEA to determine if those genes with top ranking NtcA sites were enriched among differentially expressed genes during N starvation. Analyses were done for both strains and differential expression was defined according to a gene's signed 1-P value, where the P is the CyberT P-value. <a href="GSEA/index.html">Click here for details</a>.    
    <p>References:<br>
    Baldi, P., Long, A.D. (2001) <a href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&db=pubmed&dopt=Abstract&list_uids=11395427&query_hl=3&itool=pubmed_docsum">A Bayesian framework for the analysis of microarray expression data: regularized t -test and statistical inferences of gene changes.</a> Bioinformatics. 2001 Jun;17(6):509-19.</p>
    <p>Subramanian, A., Tamayo, P., Mootha, VK., Mukherjee, S., Ebert, BL., Gillette, MA., Paulovich, A., Pomeroy, SL., Golub, Tr., Lander, ES., Mesirov, JP. (2005). <a href="http://www.pubmedcentral.gov/articlerender.fcgi?tool=pubmed&pubmedid=16199517">Gene set enrichment analysis:  a knowledge-based approach for interpreting genome-wide Expression Profiles.</a> Proceedings of the National Academy of Sciences. 102(43): 5545-5550.</p>
    <br>
    <a name="Operons"></a>
    <h2>Validation of operon predictions: correlations with gene expression patterns</h2>
    <p>The goal of these analyses was to determine if genes within putative operons were correlated in their expression patterns relative to genes in different operons. To this end, we used the expression patterns of all genes during the N starvation time series to evaluate the genome-wide MED4 and MIT9313 operon predictions used in the NtcA binding site analyses (from Su <i>et al.</i>, 2005). </p> 
    <p>For MED4, we found that the expression of genes in operon position 1 were significantly correlated with those of genes in the same operon at positions 2, 3, 4, and 5.  In MIT9313, the expression of genes in operon position 1 and position 2 were highly correlated. MIT9313 operon positions 1 vs. 3 and 1 vs. 4 were found to have elevated correlations relative to chance (p<.05), but did not survive a Bonferroni correction for multiple hypotheses. Here is a complete description of (<a href="Operons/operons.pdf">our methods and results</a>) in .pdf file format.</p>
    <br> 
    <a name="Motifs"></a> 
    <h2>Regulatory Motif Finding</h2>
    <p>Transcriptonally co-regulated genes often share an upstream sequence motif that acts as the binding site for a transcription factor. Our goal was to identify candidate transcription factor binding sites upstream of <i>Prochlorococcus</i> genes that were co-expressed during N stress. We found statistically over-represented motifs upstream of genes in specific K-means clusters.</p> 
<p>The upstream sequences that were searched for motifs were defined as 100 bp upstram of each putative MED4 or MIT9313 operon. Operons defined by the rules in Su et al., (2005), namely that tandem genes in an operon are transcribed in the same direction and are separated by <45 bp. Motif discovery analyses were done using Alignace  
run from this <a href="http://atlas.med.harvard.edu/cgi-bin/alignace.pl">html interface</a>.
 Sequence logos were generated with <a href="http://weblogo.berkeley.edu/logo.cgi">Weblogo<a/>.</p>
    <p>Motif discovery using upstream sequences from genes in specific K-means clusters:
	<ul>
	<li><a href="Motifs/medupmotifsresults.html">MED4 up-regulated K-means clusters 1 and 2</a>.</li>
	<li><a href="Motifs/mitupmotifsresults.html">MIT9313 up-regulated K-means clusters 1 and 2</a>.</li>
	<li><a href="Motifs/meddownmotifsresults.html">MED4 down-regulated K-means clusters 7 and 9</a>.</li>
	</ul></p>
    <p>In addition, we analysed all the above motifs to find motifs that were <a href="Motifs/commonmotif.html">similar in both MED4 and MIT9313</a>.</p>

<p>A complete description of the methods and results for the searches upstream of MED4 and MIT9313 up-regulated clusters in .pdf file format is <a href="./Motifs/ntcaMotifs.pdf">Here</a>.</p>

<p>A complete description of the methods and results for the searches upstream of down-regulated MED4 clusters in .pdf file format is <a href="./Motifs/medRepressed.pdf">Here</a>.</p>

    <p>References:<br>
    Hughes, JD, Estep, PW, Tavazoie S, & GM Church. <a href="http://www.sciencedirect.com/science?_ob=ArticleURL&_udi=B6WK7-45F51XV-PF&_coverDate=03%2F10%2F2000&_alid=363023854&_rdoc=1&_fmt=&_orig=search&_qd=1&_cdi=6899&_sort=d&view=c&_acct=C000050221&_version=1&_urlVersion=0&_userid=10&md5=658d91156a671304b9f2544a81554679">Computational identification of cis-regulatory elements associated with groups of functionally related genes in Saccharomyces cerevisiae.</a> Journal of Molecular Biology (2000) Mar 10;296(5):1205-14.</p> 
    <p>Roth, FR, Hughes, JD, Estep, PE & GM Church. <a href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&db=pubmed&dopt=Abstract&list_uids=9788350&query_hl=1&itool=pubmed_docsum">Finding DNA Regulatory Motifs within Unaligned Non-Coding Sequences Clustered by Whole-Genome mRNA Quantitation.</a> Nature Biotechnology 1998 Oct;16(10):939-45.</p>
    <br />
    <hr>
    Latest Update: 1-May-2006 by Andy Tolonen.
</body>
</html>
